[shaunlxw] iP

shaunlxw opened the pull request:
# Shon Bot
> "Java is so fun!" -me probably

Shon Bot frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://github.com/shaunlxw/ip/releases/tag/v0.1).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:
- [X] Managing Tasks
- [X] Managing Deadlines
- [X] Managing Events
- [ ] Reminders (Coming soon?)
- [ ] GUI (Coming soon?)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Shon {
    public static void main(String[] args) {
        new Shon.run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

shaunlxw committed:
Rename, Greet, Exit

shaunlxw committed:
Echo

shaunlxw committed:
Add, List

shaunlxw committed:
Mark as Done

- Added Task Class
- Changed Todo from List of String into List of Task
- Parse the input string from std input to read action and index separately

shaunlxw committed:
ToDos, Events, Deadlines

- Made Task class abstract
- Added Todo, Deadline, and Event class as subclasses of Task

shaunlxw committed:
Automated Text UI Testing

- Added test inputs using example from Level-4

shaunlxw committed:
Handle Errors

- Added ParameterException class
- Packaged each command into methods that throws exceptions
- Shows different text on "list" command with empty list

shaunlxw committed:
Delete

- Moved the line from class field to variable in method

shaunlxw committed:
Use Enums

-Added Action Enums
-Added CommandException class

shaunlxw committed:
feat: load initial TodoList from data file

Changes:
* add FileManager class
* change all Task subclasses constructors to support setting isDone
* overload addTodo, addDeadline and addEvent in TodoList to support creating
Task with isDone
* update gitignore for /data

shaunlxw committed:
feat: save list data after any list changes

Changes:
* add format data for each subclass of Task
* add updateData function for FileManager

shaunlxw committed:
Merge branch 'branch-Level-7'

shaunlxw committed:
Change to use LocalDateTime instead of String for date/time

* include simple exception handling for datetime parsing

shaunlxw committed:
Merge branch 'branch-Level-8'

shaunlxw committed:
Refactor code to extract out related code

* Add command subclasses with abstract class Command and AddTaskCommand
* Create Parser and Ui classes
* Rename FileManager to Storage
* Rename TodoList to TaskList

shaunlxw committed:
Merge branch 'branch-A-MoreOOP'

shaunlxw committed:
Organize classes into packages

* Change description access to protected in AddTaskCommand
* Delete getDescription() method in AddTaskCommand
* Use this.description in AddTodoCommand, AddDeadlineCommand, AddEventCommand

shaunlxw committed:
Merge branch 'branch-A-Packages'

shaunlxw committed:
Merge branch 'add-gradle-support'

shaunlxw committed:
Configure project to use gradle

shaunlxw committed:
Merge branch 'branch-A-Gradle'

shaunlxw committed:
Add JUnit test for TaskList, Todo, and Event class

* Change access of TaskList::add to private
* Reordered "public abstract" in Task class
* Change access of Task::getStatusIcon to private
* Change access of Task::formatTask to protected

shaunlxw committed:
Merge branch 'branch-A-JUnit'

shaunlxw committed:
bug fix: input validation for Parser

* Fix Parser::getDescription empty input check
* Fix Parser::getFrom empty input check
* Fix output bug that occured from code refactoring during "A-Packages" commit

shaunlxw committed:
Merge branch 'bug-fix'

shaunlxw committed:
Configure build.gradle for shadowJar

shaunlxw committed:
Merge branch 'branch-A-Jar'

shaunlxw committed:
Add JavaDoc to all classes and non-trivial methods.

* Change access of AddTaskCommand to package-private
* Remove redundant import statement in Todo

shaunlxw committed:
Refactor code to comply with coding standard

* Rename list to tasks in TaskList
* Rename getList to getTasks in TaskList
* Add white space around operators
* Rename local variable output to outputs in TaskList

shaunlxw committed:
feat: find tasks by keyword

* Create new FindCommand class
* Add new Action: FIND
* Add new method getKeyword in Parser
* Add new method filterByKeyword in TaskList
* Add new method hasKeyword in Task
* Add JavaDoc for all methods related to find

shaunlxw committed:
Merge branch 'branch-A-JavaDoc'

shaunlxw committed:
Merge branch 'branch-A-CodingStandard'

* Resolved conflicts in TaskList and MarkCommand

shaunlxw committed:
Merge branch 'branch-Level-9'

shaunlxw committed:
bug fix: rename list to tasks in filterByKeyword

shaunlxw committed:
Merge branch 'bug-fix'

shaunlxw committed:
Parser class: rename split to parameters

shaunlxw committed:
Merge branch 'branch-A-CodingStandard'

Darkarche3 commented on src/main/java/shon/Parser.java:
```diff
@@ -0,0 +1,225 @@
+    public static Command parse(String input) throws CommandException, ParameterException {
+        Action action = Parser.getAction(input);
+        switch (action) {
+        case LIST:
```
consider adding break for each case as shown in the coding standards

shaunlxw replied:
Hm I was considering if it was needed since all my cases have a return statement. Putting a break after the return statement seems quite pointless and the IDE flags the break statement as an unreachable statement and recommends it to be removed.

Darkarche3 commented on src/main/java/shon/Parser.java:
```diff
@@ -0,0 +1,225 @@
+     * @throws ParameterException If the input is a deadline command but <code>"/by"</code> is not found in the input,
+     * or if the given <code>/by</code> datetime is empty.
+     */
+    private static String getBy(String input) throws ParameterException {
```
Good that you followed the naming convention for method name

Darkarche3 commented on src/main/java/shon/Shon.java:
```diff
@@ -0,0 +1,54 @@
+package shon;
```
good that you renamed from Duke to your bot's name Shon

Darkarche3 commented on src/main/java/shon/Storage.java:
```diff
@@ -0,0 +1,111 @@
+    private void addTodo(String data, TaskList list) {
+        String[] d = data.split(" \\| ", 3);
+        String isDoneStatus = d[1], description = d[2];
+        boolean isDone = isDoneStatus.equals("1");
```
good that you remember naming conventions for boolean variables!

Darkarche3 commented on src/main/java/shon/task/Deadline.java:
```diff
@@ -0,0 +1,62 @@
+    /** The by datetime of the task */
+    private LocalDateTime by;
+    /** The formatter for reading/storing the by datetime */
+    private static DateTimeFormatter ioFormatter = DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm");
```
Curious if you really need static

shaunlxw replied:
I included static for the formatter so that all deadlines will only use the one common formatter so that the date format of all Deadline objects are the same. If not set as static, each deadline could technically have it's own formatter, and the output might be different for different Deadline objects, which is not ideal.

Darkarche3 reviewed (commented):
Good job on your code and good organisation.

shaunlxw committed:
Configure project with checkstyle

shaunlxw committed:
Merge branch 'branch-A-CheckStyle'

TarinPairor commented on src/main/java/shon/Parser.java:
```diff
@@ -85,15 +85,15 @@ private static Action getAction(String input) throws CommandException {
     private static int getIdx(String input) throws ParameterException {
-        String[] split = input.split(" ");
-        if (split.length == 1) {
+        String[] parameters = input.split(" ");
```
Very good usage of variable names, a lot of people I have seen use names like `arr` but this is better for code readability!

TarinPairor commented on src/main/java/shon/Shon.java:
```diff
@@ -7,22 +7,32 @@
 public class Shon {
-
     private Ui ui;
     private TaskList list;
```
Maybe `taskList` would be better as a more descriptive variable name?

TarinPairor commented on src/main/java/shon/Shon.java:
```diff
@@ -7,22 +7,32 @@
+ */
 public class Shon {
-
     private Ui ui;
```
Just curious, why not make `Ui` static like the `Parser` class? Was it a purposeful design choice?

TarinPairor reviewed (commented):
Overall, too well written. Very good OOP abstraction and concise, readable code!

TarinPairor commented on src/main/java/shon/Parser.java:
```diff
@@ -13,12 +13,23 @@
 public class Parser {
-
+    /** The allowed set of actions that the user can do */
     private enum Action {
```
Smart usage of Enums.

shaunlxw committed:
Add basic GUI using JavaFX

shaunlxw committed:
Add implemention for getResponse

shaunlxw committed:
Merge branch 'branch-Level-10'

shaunlxw committed:
Refactor to use FXML

* Remove dead code
* Change to adhere checkstyle

shaunlxw committed:
Merge branch 'branch-Level-10'

shaunlxw committed:
Config build.gradle for entry point for JavaFX

shaunlxw committed:
Merge branch 'branch-Level-10'

shaunlxw committed:
Add assertions in Parser, Shon and Storage classes

The code does not have assertion statements.

Assertion statements should be used to define assumptions about the
code and verify them during runtime. Assertion failure can help
identify bugs in the code.

Add assertion statements to Parser, Shon and Storage classes. Let's:
* add null check for Action in Parser#parse
* check for existence of storage directory and storage file after
initialisation
* check the correct letter is used to represent tasks in add methods
of Storage class
* check that storage and tasklist is initialised properly in Shon.

The above checks are done as assertions and not Exceptions as the
failure of above is caused by programmer error.

shaunlxw committed:
Refactor code to improve readability

The tasklist is formatted into storage format before being passed into
updateData() method.

This is not very good abstraction as it exposes the storage format to the Shon
class.

Let's handle the formatting only within the storage class. Pass the whole
tasklist into the updateData() method.

Hides the storage format from other classes.

Other changes:
* add JavaDoc comments for non-trivial public and private methods
* add comments in code and separate into logical blocks
* Parser class: abstract out checkContains and checkNotEmptyAfter
* Shon class: rename list to tasks to better represent TaskList
* TaskList class: use for-each loop to iterate through List
* TaskList class: rename Task enum to TaskType to prevent confusion with
the Task class

shaunlxw committed:
TaskList class: refactor code to use streams

shaunlxw committed:
Merge pull request #2 from shaunlxw/branch-A-Assertions

Add assertions in Parser, Shon and Storage classes

shaunlxw committed:
Merge branch 'master' into branch-A-CodeQuality

shaunlxw committed:
Merge branch 'master' into branch-A-Streams

shaunlxw committed:
Merge pull request #3 from shaunlxw/branch-A-CodeQuality

Refactor code to improve code quality and readability

shaunlxw committed:
Merge branch 'master' into branch-A-Streams

shaunlxw committed:
Merge pull request #4 from shaunlxw/branch-A-Streams

TaskList class: refactor to use Streams

shaunlxw committed:
feat: support managing notes

Command only supports task related commands, Command::execute takes in a
TaskList as a parameter.

To support note related commands, Command needs to be more general.

Let's:
* change execute to take in no parameters
* declare note related classes and commands
* allow Parser class to pass in notelist or tasklist as needed
* extend storage to store the data for the notelist too
* Restructure the package to accomodate commands for tasks and notes

shaunlxw committed:
Merge branch 'branch-D-Notes'

shaunlxw committed:
Improve GUI

Let's:
* change bot and user image to circle
* add padding to text
* display error message in red text

shaunlxw committed:
Merge branch 'branch-A-BetterGui'

shaunlxw committed:
Add help command to show available commands

shaunlxw committed:
Merge branch 'branch-Help'

shaunlxw committed:
Improve GUI

Let's:
* reduce radius for image
* remove top padding for label

shaunlxw committed:
Merge branch 'branch-A-BetterGui'

shaunlxw committed:
HelpCommand class: display commands as list

shaunlxw committed:
Merge branch 'branch-Help'

shaunlxw committed:
Add a User Guide for the project

* Replace botImage and userImage
* Include Ui.png

shaunlxw committed:
Merge branch 'branch-A-UserGuide'

shaunlxw committed:
Fix style errors and add comments

shaunlxw committed:
Merge branch 'branch-A-CodeQuality'

shaunlxw committed:
Add text bubble for messages

shaunlxw committed:
Create new component for ShonDialogBox

* Change color of ShonDialogBox to be grey

shaunlxw committed:
Merge branch 'branch-A-BetterGui'

shaunlxw committed:
Replace Ui.png for new GUI

shaunlxw committed:
Merge branch 'branch-A-UserGuide'

shaunlxw committed:
Restyle User Guide

* Remove emojis that do not work

shaunlxw committed:
Merge branch 'branch-A-UserGuide'

shaunlxw committed:
Update tests to reflect output

shaunlxw committed:
Merge branch 'branch-A-JUnit'

shaunlxw committed:
Adhere to checkstyle rules for test code

shaunlxw committed:
Merge branch 'branch-A-CodeQuality'

shaunlxw committed:
Add more error handling

Handles:
* invalid datetime for event, when from datetime is after or same as the To
datetime
* invalid data format in tasks storage file

shaunlxw committed:
Merge branch 'branch-A-MoreErrorHandling'

shaunlxw committed:
Update test case to adhere to valid datetime

shaunlxw committed:
Merge branch 'branch-A-MoreErrorHandling'

shaunlxw committed:
Add more test cases

Changes:
* moved TaskListTest into task package to reflect hierarchy in main
* add test case for from datetime after to datetime.

shaunlxw committed:
Merge branch 'branch-A-MoreTesting'

shaunlxw committed:
Add section to edit data file

shaunlxw committed:
Merge branch 'branch-A-UserGuide'

shaunlxw committed:
Correct coding style violations

shaunlxw committed:
Merge branch 'branch-A-CodeQuality'
